### ISSUE TITLE
Align plots with hyperparameters 

### ISSUE THREAD

**Issue by gaspardbb** [aimhubio/aim#491]
# Problem 

Currently, we can plot metrics on the y-axis with `step`, `epoch`, `time`, or another metric. It would be really useful to be able to choose any column for each axis, be it:
* hyperparameters passed with `set_params`;
* context passed with `track`.

## Example 

It is common to plot e.g. `test_loss` function on a regularization parameter `penalty`, to see how sensible is the test loss on the choice of regularization. Or to plot `test_loss` function of `n_samples` to have a look at the sample complexity. 

## Solution

If the table was a big Pandas' `DataFrame`, one would do: `plt.plot(df['penalty'], df['test_loss'])`. A similar trick with Aim would be awesome.

**Comment by SGevorg** [comment]
@gorarakelyan @roubkar,  @newffy is interested to contribute to this issue. 
Potentially a scatterplot solution using our [scatterplot](https://github.com/aimhubio/aim/tree/main/aim/web/ui_v2/src/components/Charts/ScatterPlot) component? 
What needs to be done to have the scatterplot fully implemented and integrated?

**Comment by roubkar** [comment]
> @gorarakelyan @roubkar, @newffy is interested to contribute to this issue.
> Potentially a scatterplot solution using our [scatterplot](https://github.com/aimhubio/aim/tree/main/aim/web/ui_v2/src/components/Charts/ScatterPlot) component?
> What needs to be done to have the scatterplot fully implemented and integrated?

We probably need to refactor our scatterplot component at first, using our current charting helper functions.
Maybe this would be a new `explorer app`?

**Comment by SGevorg** [comment]
> We probably need to refactor our scatterplot component at first, using our current charting helper functions.
> Maybe this would be a new `explorer app`?

@roubkar yea. we could add it as a new explorer app in the beginning and modify as per feedback? 
re refactor: how big of a work is it?

**Comment by roubkar** [comment]
> > We probably need to refactor our scatterplot component at first, using our current charting helper functions.
> > Maybe this would be a new `explorer app`?
> 
> @roubkar yea. we could add it as a new explorer app in the beginning and modify as per feedback?
> re refactor: how big of a work is it?

Need to discuss it with the team, @rubenaprikyan and @KaroMourad can you take a look at the current scatterplot implementation?

**Comment by gorarakelyan** [comment]
@newffy are you looking for this kind of visualization? 

![image](https://user-images.githubusercontent.com/13848158/132657006-b3a980af-8782-4979-942f-e8b620fc8e7f.png)

**Comment by newffy** [comment]
> @newffy are you looking for this kind of visualization?
> 
> ![image](https://user-images.githubusercontent.com/13848158/132657006-b3a980af-8782-4979-942f-e8b620fc8e7f.png)

Yes, kind of. It would be great to have an option to connect point with lines (line plot), it can be more readable in some circumstances

**Comment by gorarakelyan** [comment]
@newffy awesome, thanks for the details, @roubkar will share the steps of implementation shortly.

**Comment by roubkar** [comment]
Hi @newffy,

Sorry for the late response.
As already mentioned currently we are in the process of developing v3.

Aim's client-side application is a client-side rendered single page application built with `TypeScript` and `React.js`. `D3.js` library is used for charting.
For markup and styling we're using `Material-UI` React component library and `SCSS`.

Aim's UI platform consists of static pages (home page, tags page, bookmarks page, run details page) and explorer pages.
Currently, we are working on 3 explorer apps: `Runs Explorer`, `Metrics Explorer`, `Params Explorer`.

Each explorer app consists of a model(which is responsible for data fetching, data processing, state management) and a set of reusable components for query construction and data visualization.

Implemented reusable components are: `Select Forms`, `Table`, `ChartPanel`(as a multiple charts wrapper), `LineChart` and `HighPlot`.
We also have an early and basic implementation of `ScatterPlot` but it should be revisited as we have created a library of utility functions for easy chart creation.

To resolve this issue (Ability to align metric last value data with hyperparameter) we can introduce a new app, let's say `Metric-Param Correlations Explorer`.
The model of that app can be very much like `Params Explorer`'s as it can use same data for visualization: https://github.com/aimhubio/aim/blob/main/aim/web/ui_v2/src/services/models/params/paramsAppModel.ts.

The set of components can include: 
- A new `Select Form` which will consist of two select dropdowns(metric, param) and text input for the runs search query
  (Align [`metric`] by [`param`] for the given query)
- `ChartPanel` with a new type for rendering `ScatterPlot` with a custom controls bar
- `Table`

Aim's client side is built using container pattern: all our pages have a wrapper container component that takes care of the functionality(subscribes to model's state changes) and a presentational component that only returns JSX elements(markup).
See source code for ParamsContainer and Params components here: https://github.com/aimhubio/aim/tree/main/aim/web/ui_v2/src/pages/Params.

The implementation of ScatterPlot will be a lot like LineChart's: https://github.com/aimhubio/aim/blob/main/aim/web/ui_v2/src/components/LineChart/LineChart.tsx
Just instead of `drawLines` I think we'll need to add a new function `drawCircles` which will add an svg circle element for each data point.
You can find all our D3 helper functions here: https://github.com/aimhubio/aim/tree/main/aim/web/ui_v2/src/utils/d3

We can have a toggle button in ChartPanel Controls section which will pass a boolean prop(`connectWithLine`) to `ScatterPlot` as you have mentioned that it would be great to have an option to connect points with line.
Based on `connectWithLine` props value we can have an additional function which will draw a single line by passing all data points to `drawLines` function.

Hope my message will help you to get insights on Aim's UI architecture.

**Comment by gorarakelyan** [comment]
closing the issue, as Scatters Explorer has been available since Aim v3.4
